packages/fa-icon-chooser/src/components/fa-icon/fa-icon.tsx

```diff
@@ -126,33 +124,103 @@ export class FaIcon {
-      })
-      .catch(e => {
-        console.error(`${CONSOLE_MESSAGE_PREFIX}: fa-icon: failed when using 'getUrlText' to fetch icon`, e, this);
-      })
-      .finally(() => {
-        this.loading = false;
-      });
+    const processIconDefinition = (response: string) => {
+      let iconDefinition;
+
+      try {
+        if (iconUrl.endsWith('.json')) {
+          // For FA7+, parse the JSON response directly
+          const parsed = JSON.parse(response);
+
+          const iconData = Array.isArray(parsed) ? parsed : parsed.icon;
+
+          if (!iconData || !Array.isArray(iconData) || iconData.length < 5) {
+            throw new Error('Invalid icon data structure in JSON response');
+          }
+
+          iconDefinition = {
+            prefix: this.stylePrefix,
+            iconName: this.name,
+            icon: iconData,
+          };
+        } else {
+          // For FA6, parse the SVG response
+          try {
+            const svgData = parseSvgText(response);
+
+            if (!Array.isArray(svgData)) {
+              throw new Error('SVG parser did not return an array');
+            }
+
+            if (svgData.length < 5) {
+              throw new Error(`SVG data array has ${svgData.length} elements, expected 5 or more`);
+            }
+
+            // Validate individual elements
+            const [width, height, ligatures, _unicode, pathData] = svgData;
+            if (typeof width !== 'number' || typeof height !== 'number') {
+              throw new Error('Invalid width/height in SVG data');
+            }
+
+            if (!Array.isArray(ligatures)) {
+              throw new Error('Invalid ligatures array in SVG data');
+            }
+
+            if (typeof pathData !== 'string' && !Array.isArray(pathData)) {
+              throw new Error('Invalid path data in SVG data');
+            }
+
+            iconDefinition = {
+              iconName: this.name,
+              prefix: this.stylePrefix,
+              icon: svgData,
+            };
+          } catch (e) {
+            console.error(`${CONSOLE_MESSAGE_PREFIX}: SVG parsing error:`, e);
+            throw new Error(`SVG parsing error: ${e.message}`);
+          }
+        }
+
+        library.add(iconDefinition);
+        this.setIconDefinition(iconDefinition);
+      } catch (e) {
+        console.error(`${CONSOLE_MESSAGE_PREFIX}: fa-icon: failed to process icon definition:`, e);
+        throw e;
+      }
+    };
+
+    const fetchIcon = (fn: UrlTextFetcher) => {
+      this.loading = true;
+      fn(iconUrl)
+        .then(processIconDefinition)
+        .catch(e => {
+          console.error(`${CONSOLE_MESSAGE_PREFIX}: fa-icon: failed when using 'getUrlText' to fetch icon`, e, this);
+        })
+        .finally(() => {
+          this.loading = false;
+        });
+    };
+
+    if ('function' !== typeof this.getUrlText) {
+      const getUrlTextFn = Object.getOwnPropertyDescriptor(this, 'getUrlText')?.get?.();
```

Comment:
Is this thing with `getUrlTextFn` necessary? It seems extra. We haven't had it before. It seems like: if we know that `this.getUrlText` is not a function, then we can emit the error straight away.

When would it be the case that `'function' !== this.getUrlText()`, yet `'function' === typeof getUrlTextFn`?

That's the case that this block addresses. I don't yet see why it would be desirable. What do you think?

Reply:
I removed and refactored